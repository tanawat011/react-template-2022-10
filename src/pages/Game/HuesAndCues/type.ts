/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
