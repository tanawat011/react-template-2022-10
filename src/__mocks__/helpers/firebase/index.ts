export const deleteCollection = jest.fn()
export const deleteDocument = jest.fn()
export const getDocument = jest.fn()
export const getAllDocuments = jest.fn()
export const setDocument = jest.fn()
export const subscribeCollection = jest.fn()
export const updateDocument = jest.fn()
