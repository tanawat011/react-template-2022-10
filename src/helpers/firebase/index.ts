export * from './deleteCollection'
export * from './deleteDocument'
export * from './getAllDocuments'
export * from './getDocument'
export * from './setDocument'
export * from './subscribeCollection'
export * from './updateDocument'
