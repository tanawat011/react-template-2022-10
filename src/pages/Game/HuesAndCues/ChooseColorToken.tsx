import type { RoomPlayer } from './_type'

import { TwRow } from 'components/Common'
import { colorToken } from 'data/HuesAndCues'

import { CellClickable, MarkerLG } from './Common'

type Prop = {
  currentPlayer: RoomPlayer
  playersInRoom: RoomPlayer[]
  updateCurrentPlayer: (currentPlayer: RoomPlayer) => void
}

export const ChooseColorToken: React.FC<Prop> = ({
  currentPlayer,
  playersInRoom,
  updateCurrentPlayer,
}) => {
  const handleChooseColorToken = async (color: string) => {
    const _currentPlayer: RoomPlayer = {
      ...currentPlayer,
      color,
    }

    updateCurrentPlayer(_currentPlayer)
  }

  return (
    <div className='mb-2 p-3 shadow-around'>
      <div className='pl-2 mb-1 text-white font-bold select-none'>Choose Color Token</div>

      <TwRow>
        {colorToken.map((color) => {
          const _players = playersInRoom.find((p) => p.color === color)

          if (!_players) {
            return (
              <CellClickable
                key={`choose-color-token-${color}`}
                className='bg-transparent'
                onClick={() => handleChooseColorToken(color)}
              >
                <MarkerLG className={color} />
              </CellClickable>
            )
          }
        })}
      </TwRow>
    </div>
  )
}
