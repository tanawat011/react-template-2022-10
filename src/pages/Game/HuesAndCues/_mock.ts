import type { RoomPlayer } from './_type'

const PLAYER_BASE: RoomPlayer = {
  cells: [],
  uid: '1',
  name: 'Player 1',
  score: 0,
  color: '',
  number: 1,
}

const result = 'F4'
const choice = ['F4', 'A9', 'P27', 'E15']

export const MOCK: {
  ONE_PLAYER: RoomPlayer[]
  TWO_PLAYER: RoomPlayer[]
  TWO_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER: RoomPlayer[]
  THREE_PLAYER_SECOND_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_TURN: RoomPlayer[]
  HINTER_NO_CHOICE: RoomPlayer[]
  HINTER_NO_SELECT_CHOICE: RoomPlayer[]
  HINTER_SELECTED_CHOICE: RoomPlayer[]
  PLAYER_NO_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: RoomPlayer[]
} = {
  ONE_PLAYER: [
    {
      ...PLAYER_BASE,
      cells: ['D17'],
      number: 1,
      hinter: true,
      isMyTurn: true,
      color: 'bg-[#0758a5]',
    },
  ],
  TWO_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
  ],
  TWO_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
  ],
  THREE_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
      owner: true,
      score: 1,
    },
    {
      ...PLAYER_BASE,
      uid: '2',
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      uid: '3',
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_TURN: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      isMyTurn: true,
    },
  ],
  THREE_PLAYER_SECOND_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      hinter: true,
    },
  ],
  HINTER_NO_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_NO_SELECT_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_SELECTED_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
      result,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_NO_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      color: 'bg-[#117822]',
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR_START_GAME: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      result: 'F4',
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      cells: ['F4'],
    },
    {
      ...PLAYER_BASE,
      number: 3,
      cells: ['F5', 'F6'],
    },
  ],
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      isSubmitResult: true,
      countRound: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
}
