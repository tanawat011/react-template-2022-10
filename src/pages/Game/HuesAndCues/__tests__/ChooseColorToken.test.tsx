import type { RenderResult } from '@testing-library/react'

import { act, fireEvent, screen, waitFor } from '@testing-library/react'

import { renderWithProviders } from 'helpers/test'

import { ChooseColorToken } from '../ChooseColorToken'
import { MOCK } from '../_mock'

jest.mock('../Common', () => ({
  ...jest.requireActual('../Common'),
  CellClickable: (props: React.HTMLAttributes<never>) => {
    return (
      <div {...props} data-testid={props.className}>
        {props.children}
      </div>
    )
  },
}))

describe('<ChooseColorToken />', () => {
  test('match snapshot', () => {
    const updateCurrentPlayer = jest.fn()

    const { asFragment } = renderWithProviders(
      <ChooseColorToken
        playersInRoom={MOCK.PLAYER_NO_COLOR}
        currentPlayer={MOCK.PLAYER_NO_COLOR[0]}
        updateCurrentPlayer={updateCurrentPlayer}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('current player selecting color token', async () => {
    let el: RenderResult

    const updateCurrentPlayer = jest.fn()

    await act(async () => {
      el = renderWithProviders(
        <ChooseColorToken
          playersInRoom={MOCK.PLAYER_NO_COLOR}
          currentPlayer={MOCK.PLAYER_NO_COLOR[0]}
          updateCurrentPlayer={updateCurrentPlayer}
        />,
      )
    })

    const button = screen.getAllByTestId('bg-transparent')[5]
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('some players has color token', () => {
    const updateCurrentPlayer = jest.fn()

    const { asFragment } = renderWithProviders(
      <ChooseColorToken
        playersInRoom={MOCK.PLAYER_HAS_COLOR}
        currentPlayer={MOCK.PLAYER_HAS_COLOR[0]}
        updateCurrentPlayer={updateCurrentPlayer}
      />,
    )
    expect(asFragment()).toBeTruthy()
  })
})
