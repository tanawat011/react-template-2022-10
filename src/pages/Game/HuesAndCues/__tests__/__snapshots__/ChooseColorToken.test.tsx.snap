// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseColorToken /> match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  width: 0.75rem;
  height: 0.75rem;
  border-radius: 9999px;
  border-width: 1px;
  --tw-shadow: 0 25px 50px -12px rgb(0 0 0 / 0.25);
  --tw-shadow-colored: 0 25px 50px -12px var(--tw-shadow-color);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c2 {
  width: 1.5rem;
  height: 1.5rem;
}

<div
    class="mb-2 p-3 shadow-around"
  >
    <div
      class="pl-2 mb-1 text-white font-bold select-none"
    >
      Choose Color Token
    </div>
    <div
      class="c0"
    >
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#e10655]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#c00502]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#c33200]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#dac204]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#117822]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#0758a5]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#4e306f]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#858777]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#d5d3cb]"
        />
      </div>
      <div
        class="bg-transparent"
        data-testid="bg-transparent"
      >
        <div
          class="c1 c2 bg-[#18110d]"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
